src/core/include/mp-units/reference.h

```diff
@@ -53,22 +53,24 @@ namespace mp_units {
  * The following syntaxes are not allowed:
  * `2 / kmph`, `kmph * 3`, `kmph / 4`, `70 * isq::length[km] / isq:time[h]`.
  */
-template<QuantitySpec auto Q, Unit auto U>
+template<QuantitySpec Q, Unit U>
 struct reference {
-  template<auto Q2, auto U2>
+  template<typename Q2, typename U2>
   [[nodiscard]] friend consteval bool operator==(reference, reference<Q2, U2>)
   {
-    return Q == Q2 && U == U2;
+    return Q{} == Q2{} && U{} == U2{};
   }
 
   template<AssociatedUnit U2>
   [[nodiscard]] friend consteval bool operator==(reference, U2 u2)
   {
-    return Q == get_quantity_spec(u2) && U == u2;
+    return Q{} == get_quantity_spec(u2) && U{} == u2;
   }
 
-  template<auto Q2, auto U2>
-  [[nodiscard]] friend consteval reference<Q * Q2, U * U2> operator*(reference, reference<Q2, U2>)
+  template<typename Q2, typename U2>
+  [[nodiscard]] friend consteval reference<std::remove_const_t<decltype(Q{} * Q2{})>,
```

Comment:
Quantity multiplication should return by value.
If you're using this for something else, I suggest `std::type_identity_t`
or something more descriptive in a `detail` namespace.
Same for #550.

Reply:
I am not sure what you mean? The code above encapsulates the return types of multiplication of two quantity_spec and two units into the `reference<QS, U>` wrapper. Those are not `quantity` values being multiplied here.

Reply:
Oh, it was for `quantity_spec`.
Still, that should return by value.
I thought you were using `std::remove_const_t` here
to make Clang change the error output.

Reply:
No, I just don't want to end up with `reference<const isq::length, const si::metre>` which sometimes happens (but maybe not in this case).

What do you mean by return by value? The purpose of this PR is to replace NTTP with a type.

Reply:
I mean the result of `*`.

Reply:
The result of `*` on two reference objects is a reference storing the types returned from `*` on ingredients. I do not know how to code it differently.

The previous implementation read better without the need for `decltype` and `std::remove_const_t` but produced worse error messages. The user's code is not impacted by this change as the user should never instantiate the `reference` class template explicitly but use `isq::length[si::metre]` to get a proper reference instance.

Reply:
Hopefully, this issue is resolved by my last commit.

Reply:
But it fails on clang-16 :-(
@johelegp, do you have any idea how to workaround the clang-16 linking issue while still making the `reference` interfaces nicer?

Reply:
No idea, sorry.

> No, I just don't want to end up with `reference<const isq::length, const si::metre>` which sometimes happens (but maybe not in this case).

That doesn't happen here, because the argument to `reference` is the result of an `operator*`.

Reply:
It is clearly a compiler bug related to automated type deduction of function templates that prevents us from doing it in a nice way. I do not want to disable clang-16 and apple-clang-15 support because of this so I will probably revert the last commit.

Reply:
> That doesn't happen here, because the argument to reference is the result of an operator*.

Yes, we will need to carefully review all `std::remove_const_t` and `std::remove_cvref_t` while writing API docs. We will also need to ensure that we have unit tests for such cases to ensure that the `const` will not somehow magically lead into a type. Or maybe we can add constraints in class templates to ensure `const` is not provided but I am not sure if doing such checks in the official interfaces is a good idea (maybe only for debug builds?).